### PR TITLE
[SPARK-20204][SQL] remove SimpleCatalystConf and CatalystConf type alias

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is a follow-up of https://github.com/apache/spark/pull/17285 .

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @hvanhovell  @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystConf.scala [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+  val RUN_SQL_ON_FILES = buildConf("spark.sql.runSQLOnFiles")
+    .internal()
+    .doc("When true, we could use `datasource`.`path` as table in SQL query.")
+    .booleanConf
+    .createWithDefault(true)
+
+  // Use lazy val here, as the default timezone may change after `CatalystConf` is instantiated.
+  // TODO: we may need a `ConfigEntryWithDefaultFunction` so that the default value of session
+  // location timezone is always the default timezone at the time we call `getConf`.
+  lazy val SESSION_LOCAL_TIMEZONE =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystConf.scala]: cc @ueshin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystConf.scala [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+  override def clone(): CatalystConf = {
+    val result = new CatalystConf
+    getAllConfs.foreach {
+      case(k, v) => if (v ne null) result.setConfString(k, v)
+    }
+    result
+  }
+
+  // For test only
+  private[spark] def copy[T](entry: ConfigEntry[T], value: T): CatalystConf = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystConf.scala]: Most of the content in this file is copied from `SQLConf`, except the `lazy val` for session local timezone, and the 3 `copy` methods here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ConfEntries.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+import org.apache.spark.internal.config.{ConfigBuilder, ConfigEntry}
+
+object ConfEntries {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ConfEntries.scala]: copied from `object SQLConf`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 4]
```diff
@@ -1,84 +0,0 @@
...
-package org.apache.spark.sql.internal
-
-import org.apache.spark.util.Utils
-
-
-/**
- * Static SQL configuration is a cross-session, immutable Spark configuration. External users can
- * see the static sql configs via `SparkSession.conf`, but can NOT set/unset them.
- */
-object StaticSQLConf {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: moved to `CatalystConf.STATIC` and `SQLConf.STATIC`

**Comment by rxin** [comment]
To be clear, I don't think we should have two separate places to define config entries. If this is what the pr is doing, I strongly veto.

**Comment by rxin** [comment]
LGTM

**Comment by rxin** [comment]
Merging in master.

**Comment by nsyca** [comment]
After merging my local branch up to this PR, I ran some of the regression tests from a machine in the Eastern time zone and observed the following failures:

````
[info] *** 35 TESTS FAILED ***
[error] Failed: Total 3094, Failed 35, Errors 0, Passed 3059, Ignored 54
[error] Failed tests:
[error] 	org.apache.spark.sql.SQLQuerySuite
[error] 	org.apache.spark.sql.DateFunctionsSuite
[error] 	org.apache.spark.sql.SQLQueryTestSuite
[error] 	org.apache.spark.sql.execution.datasources.csv.CSVSuite
[error] 	org.apache.spark.sql.DataFrameSuite
[error] 	org.apache.spark.sql.sources.PartitionedWriteSuite
[error] 	org.apache.spark.sql.JsonFunctionsSuite
[error] 	org.apache.spark.sql.execution.datasources.json.JsonSuite
[error] 	org.apache.spark.sql.execution.datasources.parquet.ParquetQuerySuite
[error] (sql/test:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 1132 s, completed 4-Apr-2017 6:09:08 PM
````

One failure is SQLQuerySuite/SPARK-3173 Timestamp support in the parser. Once I change my machine to Pacific time zone, the failure is gone.

I am not making a statement that this PR causes the above failures. It could be something before this PR. I just want to share this information.

**Comment by rxin** [comment]
@nsyca can you look into it?

**Comment by nsyca** [comment]
I will investigate. I am searching from the last good point I merged my private branch with the master trunk and will go from there.

**Comment by cloud-fan** [comment]
@nsyca can you try marking `SQLConf.SESSION_LOCAL_TIMEZONE` as `lazy val`? I think the issue is that, once `object SQLConf` is instantiated, the default value for `SESSION_LOCAL_TIMEZONE` is determined, and later in the test we will change the timezone which can't take effect then.

**Comment by nsyca** [comment]
@dilipbiswal has narrowed down that this PR is changing the behaviour. He will continue to investigate and will post an update in the next hour or so before he calls it a day.

**Comment by dilipbiswal** [comment]
@cloud-fan @nsyca Changing to make it lazy works for the test cases i have tried. I am running the full tests now.

**Comment by dilipbiswal** [comment]
@cloud-fan @nsyca A quick update.. I ran the problematic tests and they pass with a change to move the time zone setting code to PlanTest.scala just before we create the SQLConf like following -

```scala
 // Timezone is fixed to America/Los_Angeles for those timezone sensitive tests (timestamp_*)
  TimeZone.setDefault(TimeZone.getTimeZone("America/Los_Angeles"))
  // Add Locale setting
  Locale.setDefault(Locale.US)

  protected val conf = new SQLConf().copy(SQLConf.CASE_SENSITIVE -> true)
```
Please let me know what you think ..

**Comment by gatorsmile** [comment]
`SESSION_LOCAL_TIMEZONE` sounds a static SQLConf. Do we allow users to change it at runtime? It sounds like our codes do not allow users to change/refresh it.

**Comment by cloud-fan** [comment]
also cc @ueshin , I think the default value of `SESSION_LOCAL_TIMEZONE` should always be the current timezone in the JVM. We may need something like `ConfigEntryWithDefaultFunction`, so that the default value is not fixed when the conf entry is created.

**Comment by gatorsmile** [comment]
These test cases are timezone sensitive, right? If so, the changes made by @dilipbiswal are reasonable to me.

**Comment by ueshin** [comment]
I agree with @cloud-fan's suggestion, `ConfigEntryWithDefaultFunction` approach. Hopefully the default value should be fixed when `SQLConf` instance is created with the timezone in the JVM at that moment.

**Comment by gatorsmile** [comment]
@dilipbiswal Can you please try it based on what @cloud-fan and @ueshin suggested? Does it resolve the issue you report?

**Comment by dilipbiswal** [comment]
@gatorsmile @cloud-fan @ueshin Sorry .. i was on transit from work. Sure, i will make a try. However , i wanted to understand this a bit more. In my understanding, the current problem we are trying to address is only in our test environment.  In our tests as part of the infrastructure we want to keep the timezone constant as we have hardcoded timestamp literals and in order to make row comparisions work. Today, we do that in QueryTest.scala [here](https://github.com/apache/spark/blob/f48461ab2bdb91cd00efa5a5ec4b0b2bc361e7a2/sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala#L43-46)

However, due to the change in the PR, this is happening a bit late as we have created a cloned copy of SQLConf in PlanTest.scala (super class of QueryTest.scala) in [here](https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala#L32). So in my understanding its this copy step where the configurations are used the first time. Please correct me on this.. Thats the reason, i was suggesting to move the setting of timezone up into PlanTest before the place where SQLConf is cloned.

I will study the ConfigEntryWithDefault approach in the meantime.

**Comment by nsyca** [comment]
Just add another view point to this incident.

If those failed test cases were written with the time zone fixed to a region rather than Pacific time zone, we would fail fast in the first run over the regression machines in the west coast. We should have picked the time zone like GMT+1, one that is different from the GMT used by `java.sql.Timestamp` and from the local time zone of the regression machines.

**Comment by viirya** [comment]
The `ConfigEntryWithDefaultFunction` approach looks good to me.

**Comment by dilipbiswal** [comment]
@viirya Thanks !! Actually i assumed the problem is limited only to tests. I thought changing the timezone on the fly is not a realistic scenario. Looks like it is :-) 

I will submit a pr with wenchen's suggestion.

**Comment by viirya** [comment]
The pr is at #17537.

**Pull request closed without merging**